Python/parking_lot.c

```diff
@@ -0,0 +1,470 @@
+        res = Py_PARK_TIMEOUT;
+    }
+#elif defined(USE_SEMAPHORES)
+    int err;
+    if (timeout >= 0) {
+        struct timespec ts;
+
+        _PyTime_t deadline = _PyTime_Add(_PyTime_GetSystemClock(), timeout);
+        _PyTime_AsTimespec(deadline, &ts);
+
+        err = sem_timedwait(&sema->platform_sem, &ts);
+    }
+    else {
+        err = sem_wait(&sema->platform_sem);
+    }
+    if (err == -1) {
+        err = errno;
+        if (err == EINTR) {
+            res = Py_PARK_INTR;
+        }
+        else if (err == ETIMEDOUT) {
+            res = Py_PARK_TIMEOUT;
+        }
+        else {
+            _Py_FatalErrorFormat(__func__,
+                "unexpected error from semaphore: %d",
+                err);
+        }
+    }
+    else {
+        res = Py_PARK_OK;
+    }
+#else
+    pthread_mutex_lock(&sema->mutex);
+    if (sema->counter == 0) {
+        int err;
+        if (timeout >= 0) {
+            struct timespec ts;
+
+            _PyTime_t deadline = _PyTime_Add(_PyTime_GetSystemClock(), timeout);
+            _PyTime_AsTimespec(deadline, &ts);
+
+            err = pthread_cond_timedwait(&sema->cond, &sema->mutex, &ts);
+        }
+        else {
+            err = pthread_cond_wait(&sema->cond, &sema->mutex);
+        }
+        if (err) {
+            res = Py_PARK_TIMEOUT;
+        }
+    }
+    if (sema->counter > 0) {
+        sema->counter--;
+        res = Py_PARK_OK;
+    }
+    pthread_mutex_unlock(&sema->mutex);
+#endif
+    return res;
+}
+
+int
+_PySemaphore_Wait(_PySemaphore *sema, _PyTime_t timeout, int detach)
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+    int was_attached = 0;
+    if (tstate) {
+        was_attached = (tstate->_status.active);
+        if (was_attached && detach) {
+            PyEval_ReleaseThread(tstate);
+        }
+    }
+
+    int res = _PySemaphore_PlatformWait(sema, timeout);
+
+    if (tstate) {
+        if (was_attached && detach) {
+            PyEval_AcquireThread(tstate);
+        }
+    }
+    return res;
+}
+
+void
+_PySemaphore_Wakeup(_PySemaphore *sema)
+{
+#if defined(MS_WINDOWS)
+    if (!ReleaseSemaphore(sema->platform_sem, 1, NULL)) {
+        Py_FatalError("parking_lot: ReleaseSemaphore failed");
+    }
+#elif defined(USE_SEMAPHORES)
+    int err = sem_post(&sema->platform_sem);
+    if (err != 0) {
+        Py_FatalError("parking_lot: sem_post failed");
+    }
+#else
+    pthread_mutex_lock(&sema->mutex);
+    sema->counter++;
+    pthread_cond_signal(&sema->cond);
+    pthread_mutex_unlock(&sema->mutex);
+#endif
+}
+
+_PySemaphore *
+_PySemaphore_Alloc(void)
+{
+    // Make sure we have a valid thread_data. We need to acquire
+    // some locks before we have a fully initialized PyThreadState.
+    _PyParkingLot_InitThread();
+
+    ThreadData *this_thread = thread_data;
+    if (this_thread->depth >= MAX_SEMA_DEPTH) {
+        Py_FatalError("_PySemaphore_Alloc(): too many calls");
+    }
+    return &this_thread->semas[this_thread->depth++];
```

Comment:
With a statically allocated `thread_data`:

```suggestion
    if (thread_data.depth >= MAX_SEMA_DEPTH) {
        Py_FatalError("_PySemaphore_Alloc(): too many calls");
    }
    return &thread_data.semas[thread_data.depth++];
```

Reply:
I've removed `thread_data`.